libgnucash/app-utils/gnc-option-impl.hpp

```diff
@@ -0,0 +1,1066 @@
+    {
+        auto iter = std::find_if(m_choices.begin(), m_choices.end(),
+                              [key](auto choice) {
+                                  return std::get<0>(choice) == key; });
+        if (iter != m_choices.end())
+            return iter - m_choices.begin();
+        else
+            return size_t_max;
+
+    }
+    GncOptionUIType m_ui_type;
+    GncMultichoiceOptionIndexVec m_value;
+    GncMultichoiceOptionIndexVec m_default_value;
+    GncMultichoiceOptionChoices m_choices;
+    static const std::string c_empty_string;
+    static const std::string c_list_string;
+};
+
+template<> inline std::ostream&
+operator<< <GncOptionMultichoiceValue>(std::ostream& oss,
+                                       const GncOptionMultichoiceValue& opt)
+{
+    auto vec{opt.get_multiple()};
+    bool first{true};
+    for (auto index : vec)
+    {
+        if (first)
+            first = false;
+        else
+            oss << " ";
+        oss << opt.permissible_value(index);
+    }
+    return oss;
+}
+
+template<> inline std::istream&
+operator>> <GncOptionMultichoiceValue>(std::istream& iss,
+                                       GncOptionMultichoiceValue& opt)
+{
+    GncMultichoiceOptionIndexVec values;
+    while (true)
+    {
+        std::string str;
+        std::getline(iss, str, ' ');
+        if (!str.empty())
+        {
+            auto index = opt.permissible_value_index(str.c_str());
+            if (index != size_t_max)
+                values.push_back(index);
+            else
+            {
+                std::string err = str + " is not one of ";
+                err += opt.m_name;
+                err += "'s permissible values.";
+                throw std::invalid_argument(err);
+            }
+        }
+        else
+            break;
+    }
+    opt.set_multiple(values);
+    iss.clear();
+    return iss;
+}
+
+
+using GncOptionAccountList = std::vector<const Account*>;
+
+using GncOptionAccountTypeList = std::vector<GNCAccountType>;
+
+/** @class GncOptionAccountListValue
+ *
+ * Set one or more accounts on which to report, optionally restricted to certain
+ * account types. Many calls to make-account-list-option will pass a get-default
+ * function that retrieves all of the accounts of a list of types.
+ *
+ * Some reports (examples/daily-reports.scm and standard/ account-piechart.scm,
+ * advanced-portfolio.scm, category-barchart.scm, net-charts.scm, and
+ * portfolio.scm) also provide a validator that rejects accounts that don't meet
+ * an account-type criterion.
+ *
+ * There are two types of option, account-list which permits more than one
+ * account selection and account-sel, which doesn't.
+ *
+
+ */
+
+class GncOptionAccountListValue : public OptionClassifier
+{
+public:
+    GncOptionAccountListValue(const char* section, const char* name,
+                          const char* key, const char* doc_string,
+                          GncOptionUIType ui_type, bool multi=true) :
```

Comment:
Would it make sense to set a default ui_type here as you did in earlier, simpler option classes ?

Reply:
That possible only for the one constructor that takes neither a default value nor a validation list, and we don't actually use that one.

Reply:
It's not immediately obvious to me why we can't specify a default ui_type. Can you elaborate on that  a bit ?

Reply:
> It's not immediately obvious to me why we can't specify a default ui_type. Can you elaborate on that a bit ?

Looking further, I presume the reason is that parameters with optional defaults have to come after parameters without defaults. And so if we want defaults for `ui_type` it'd have to move further down the parameter list of these constructors. And that would mean overhauling the whole api...

I can't really estimate at this point whether that's worth it.
On the one hand from a user of this internal api's point of view having to explicitly specify a ui_type for options that normally always use a pre-defined ui_type is a bit pointless. On the other hand I do get gnc-option-impl is internal implementation so this is not really impacting consumers of the public option api. So extra effort to get there may not really be worth it.

Reply:
Right. I prefer to keep the basic order of the constructor variables the same. They all need a ui_type parameter so for constructors that need additional parameters those go after the ui_type and that prevents those constructors from having a default ui_type. Note too that any option type can be created as an internal option so the UI type has to be overridable.

Reply:
A fair design choice, let's keep it as you implemented it.